kratos/tests/cpp_tests/utilities/test_volume_inside_voxel_utility.cpp

```diff
@@ -0,0 +1,748 @@
+//    |  /           |
+//    ' /   __| _` | __|  _ \   __|
+//    . \  |   (   | |   (   |\__ `
+//   _|\_\_|  \__,_|\__|\___/ ____/
+//                   Multi-Physics
+//
+//  License:     BSD License
+//               Kratos default license: kratos/license.txt
+//
+//  Main authors:    Ariadna Cortes
+//
+//
+
+// Project includes
+#include "containers/model.h"
+#include "includes/element.h"
+#include "geometries/triangle_3d_3.h"
+#include "includes/checks.h"
+#include "testing/testing.h"
+#include "utilities/volume_inside_voxel_utility.h"
+
+namespace Kratos {
+namespace Testing {
+namespace {
+
+    typedef Node<3> NodeType;
+    typedef Node<3>::Pointer NodePtrType;
+    typedef Geometry<NodeType> GeometryType;
+    typedef GeometryType::Pointer GeometryPtrType;
+    typedef GeometryType::GeometriesArrayType GeometryArrayType;
+    typedef GeometryType::PointsArrayType PointsArrayType;
+
+    /*this functions are creating a model for each element used in the test in order to avoid trouble with repeated Ids.
+    This is not optimal but since it is just a test (in real cases, funcions will be used with well-defined models)...
+    */
+    GeometryPtrType GenerateHexahedra3D8(const std::vector<double>& rDistances) 
```

Comment:
I would separate the `GenerateHex `and `SetDistance`. The reason is that you create the voxel once and you only change the distances afterward. In your implementation you are recreating the same nodes several times which is returning you the previously created ones and changing their distances would result in changing the original one distances.

Reply:
Okey done!